### ISSUE TITLE
[Feature] Dev / hot-reload debug mode: skip weight loading and re-run graph capture + serving after code changes

### ISSUE THREAD

**Issue by ch-wan** [sgl-project/sglang#31021]
> **Note:** This feature request was generated by an AI agent based on my rough idea. I have not carefully reviewed it.

### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

SGLang server startup roughly has three expensive phases:

1. **Weight loading** — load model weights from disk / HF into GPU memory
2. **Graph capture** — CUDA graph (and related) capture / warmup
3. **Model serving** — enter the scheduler event loop and serve requests

When developing or debugging **serving-time** behavior (NaNs, hangs, numerical correctness, scheduler logic, attention backends, sampling, etc.), we typically:

1. Edit Python code
2. Kill the process
3. Restart the server and wait through **all three** phases again
4. Reproduce the issue

In most of these loops, **weights did not change**. Weight loading is often the dominant cost (especially for large models), so the current edit–restart–reproduce cycle is much slower than it needs to be.

It would greatly improve developer productivity if SGLang offered a **debug / hot-reload mode** that keeps weights (and optionally other heavy state) alive, and only redoes the work that actually depends on the code change—typically **module reload → re-init / recapture graphs → resume serving**.

### Proposed idea

Two complementary directions (either or both would help):

#### 1. Fault-tolerant scheduler event loop (dev-only)

In debug mode, wrap the scheduler event loop (or per-request / per-batch handling) in a try/except so that:

- Serving-time exceptions (e.g. NaN checks, assertion failures, Python bugs) do **not** always tear down the whole process
- Errors are logged with rich context
- Optionally, the server stays up so the developer can fix code and trigger a reload without restarting from weight load

This is mainly about **surviving failures** during iterative debugging, not about production resilience.

#### 2. Hot-reload API: reload Python modules + recapture, without reloading weights

After editing source files, send a control request (HTTP admin endpoint, or internal IPC to the scheduler / `ModelRunner`), roughly:

1. Drain or pause in-flight requests (or reject new ones during reload)
2. `importlib.reload` (or equivalent) for selected modules
3. Re-bind model methods / re-init components that depend on reloaded code (attention backends, sampling, custom ops wrappers, etc. as needed)
4. **Recapture CUDA graphs** (and related warmup)
5. Resume the event loop / model serving

**Do not** re-run full weight loading unless weights or weight layout actually changed.

A sketch of the intended workflow:

```text
# Terminal 1: start once with large model (pay weight load + first graph capture)
python -m sglang.launch_server --model <large-model> --enable-dev-reload ...

# Edit sglang Python sources...

# Terminal 2: apply changes without full restart
curl -X POST http://localhost:30000/dev/reload   # name TBD
# server: reload modules → recapture graphs → ready again
```

### Why this is valuable

| Scenario | Today | With debug hot-reload |
|---|---|---|
| Fix NaN / correctness in forward path | Full restart (load + capture + serve) | Reload + recapture |
| Debug hang / race in scheduler | Full restart | Often reload only (if no graph change) |
| Iterate on attention backend / sampling | Full restart | Reload + recapture |
| Large models (70B+, MoE, multi-GPU) | Minutes per iteration | Seconds–tens of seconds per iteration |

This primarily targets **contributors and power users** doing local iteration, not production deployments.

### Related resources / prior art in SGLang

Some building blocks already exist and may be reusable:

- **Weight update paths** (`update_weights_from_disk` / tensor / distributed / IPC) for RL-style weight refresh without full process restart — shows in-process reconfiguration is already a thing, though focused on **weights**, not Python code.
- **`recapture_cuda_graph`** on weight-update flows — shows graph recapture can be triggered after runtime changes without a full cold start.
- **`--debug-cuda-graph` / breakable CUDA graph** — improves *serving-time* debuggability inside graphs, but still requires a full restart after code edits.
- Multi-process architecture (TokenizerManager / Scheduler / ModelRunner) — reload must be coordinated carefully across ranks/workers (TP/PP/DP/EP).

External analogies (not SGLang-specific): PyTorch Lightning / some training frameworks' "reload on change", Jupyter-style module reload, and "debug attach without reloading checkpoint" workflows in large-model training stacks.

### Suggested design sketch (for discussion)

Not a committed design—open to better approaches:

1. **Flag**: e.g. `--enable-dev-reload` / `--debug-mode` (off by default; never recommended in production).
2. **Scope of reload**:
   - Explicit module allowlist, or
   - "reload all `sglang.*` modules under a path", or
   - user-specified package list in the request body.
3. **What is preserved**: model parameters / KV pool allocation policy as much as possible; reset request state / running batches.
4. **What is rebuilt**: Python callables, CUDA graphs, possibly attention backends and other non-weight init that depends on code.
5. **Multi-GPU**: broadcast reload barrier so all ranks reload the same revision and recapture in lockstep.
6. **Safety**: clear "DEV ONLY" docs; refuse reload if unsafe state (e.g. in-flight PD disagg, mid-update weights) unless force flag is set.

### Open questions

1. Is **module-level `importlib.reload`** realistic given SGLang's multi-process layout and C++/CUDA extensions, or is a lighter approach better (e.g. only re-import selected pure-Python hooks / monkey-patch points)?
2. How far should reload go: only pure Python scheduler logic vs. model `forward` / custom ops wrappers that force graph recapture?
3. Should the first version only support **single-node TP** (or even TP=1) to reduce complexity?
4. Interaction with torch.compile, CUDA graphs, DeepGEMM / JIT kernels, and extension modules that cannot be reloaded cleanly?
5. Prefer **admin HTTP API**, **CLI signal** (e.g. SIGHUP), or both?
6. Is fault isolation in the event loop (try/except) phase 1, with hot-reload as phase 2?

### Alternatives considered

- **`--skip-server-warmup` / disable CUDA graph**: speeds startup somewhat but still reloads weights every time.
- **Smaller models for debug**: helps, but many bugs only show up on target large models / multi-GPU.
- **External process managers + weight cache on disk**: still pays graph capture and process bootstrap; does not help mid-session code iteration as cleanly.
- **Attach debugger only**: great for breakpoints, does not help "edit code → test again" without restart.

### Call to action

Looking for maintainer feedback on:

- Whether this fits the project roadmap as a **dev-only** feature
- Preferred MVP scope (event-loop resilience vs. hot-reload vs. both)
- Any hard blockers from the multi-process / CUDA graph architecture we should design around early

Happy to help prototype an MVP if the direction sounds reasonable.

**Comment by ch-wan** [comment]
Related PR: https://github.com/sgl-project/sglang/issues/27139
